### ISSUE TITLE
[Bug?] "✓ Plex Wrapped" greyed out

### ISSUE THREAD

**Issue by VP-EN** [aunefyren/wrapperr#24]
Hey

I installed wrapperr by:

1. Create a quick domain-setup, with `/etc/nginx/sites-enabled/wrapped.domain.com.conf` and enable PHP.
2. The wrapped.domain.com.conf file looks like [this](https://0bin.net/paste/YckhmVD8#3RWmfv3iG5StVGnSmetal408dvEItQCQyIiKzD29i1b)
2. `cd /var/www/wrapped.domain.com`
2. `sudo git clone https://github.com/aunefyren/wrapperr.git`
3. `sudo mv wrapperr/ html/`
4. `sudo chown -R www-data:www-data html/`
5. `sudo chmod -R 0777 /var/www/wrapped.domain.com/html/config`
6. I configured an admin login/password
6. I configured my Tautulli and confirmed it is working (green Test Connection prompt)
7. I chose the pre-cache option, 50 days
8. Waited for the cache to complete
9. Confirmed that the `cache.json` file in `/var/www/wrapped.domain.com/html/config` was created (size is about 5-6mb after pre-cache)
10. Restarting nginx, `sudo service nginx restart`
11. Trying to login using the "Sign in using Plex" prompt
12. After logging in, I get this: https://i.imgur.com/wYFJFt6.png
13. "✓ Plex Wrapped" greyed out

Can't really figure out if I did anything wrong. It worked before using the Plex OAuth login, so I'm not sure why it doesn't work now.

**Other info**
1) Tautulli is v2.8.0 (2021-12-15)
2) Using the "Sign in using Plex" prompt, gives a rather unpleasant "warning" message:
https://i.imgur.com/NigpkMQ.png

**Comment by aunefyren** [comment]
Thanks for the good explaination!

Seems to be related to the Plex Auth, cookie validation, part of the API. The button is greyed out until the cookie, created at login, is validated. Then the cookie should either be approved, making the button available, or you should be logged out and be prompted to log in again.

Is there anything in the developer console of your webbrowser? Typically found with F12 on your keyboard.

`Using the "Sign in using Plex" prompt, gives a rather unpleasant "warning" message:`

I get this on my website too, but login is working.

**Comment by VP-EN** [comment]
> Thanks for the good explaination!
> 
> Seems to be related to the Plex Auth, cookie validation, part of the API. The button is greyed out until the cookie, created at login, is validated. Then the cookie should either be approved, making the button available, or you should be logged out and be prompted to log in again.
> 
> Is there anything in the developer console of your webbrowser? Typically found with F12 on your keyboard.
> 
> `Using the "Sign in using Plex" prompt, gives a rather unpleasant "warning" message:`
> 
> I get this on my website too, but login is working.

Yeah I seem to get an error with the `validate_login_cookie.php`
https://i.imgur.com/QDggygM.png

Log:
https://0bin.net/paste/lypzHmvr#1wB69Fg9+KGv9wS8-k5r/3DhdUBLmtDUHHMXnVBYbzv

**Comment by aunefyren** [comment]
Are you able to send API/validate_login_cookie.php your cookie and see the response?

**Comment by VP-EN** [comment]
> Are you able to send API/validate_login_cookie.php your cookie and see the response? 

I have no idea how to do this. I mean, I know how to obtain the cookie. Guessing that's the cookie string from the dev console. But I have no idea how to parse it through the php file (and how to gather the response) 
Can you guide me? 🙂

**Comment by aunefyren** [comment]
It accepts Post data, not Get. You can do this using curl I think.

> -d, --data <data> Send specified data in POST request. Details provided below.

https://gist.github.com/subfuzion/08c5d85437d5d4f00e58

**Comment by VP-EN** [comment]
I tried running
```
curl --data "plex-wrapped=uEHOoq849mpPshON3yAU2yLmG96pDponZMCV2Kd6odYw9DZG" "https://wrapped.domain.com/api/validate_login_cookie.php"
```
Just gives me:
`{"error":true,"message":"Input error."}`

Am I doing it wrong? Sorry if so

**Comment by aunefyren** [comment]
Not gonna lie, I have issues sending info to API's using Curl as I have difficulties wrapping my mind around the parameters. I usually use Postman as an API tester.

The API accepts Post data, and communicates using the JSON format. So maybe the data must be a JSON string? Like this '{"cookie":"crazycookiestribg123"}'.

**Comment by VP-EN** [comment]
> Not gonna lie, I have issues sending info to API's using Curl as I have difficulties wrapping my mind around the parameters. I usually use Postman as an API tester.
> 
> The API accepts Post data, and communicates using the JSON format. So maybe the data must be a JSON string? Like this '{"cookie":"crazycookiestribg123"}'.

Okay so I tried this, taking the answer from [this stackexchange](https://stackoverflow.com/questions/7172784/how-do-i-post-json-data-with-curl):
```
curl --header "Content-Type: application/json" \
  --request POST \
  --data '{"cookie":"EGj/JNuG04eEamIFa3SXHX0o/ersRutZqOa8ulQ4PD/AEvov"}' \
  https://wrapped.domain.com/api/validate_login_cookie.php
```
which gives no response

Tried with this:
```
curl --header "Content-Type: application/json" \
  --request POST \
  --data '{"cookie":"plex-wrapped=EGj/JNuG04eEamIFa3SXHX0o/ersRutZqOa8ulQ4PD/AEvov"}' \
  https://wrapped.domain.com/api/validate_login_cookie.php
```

Which then gives:
```
{"message":"Encryption failure.","error":true,"data":[]}
```
So, finally something new :) "Encryption failure". I'm running the `https://wrapped.domain.com` behind an SSL certificate. Just used [certbot with nginx](https://certbot.eff.org/instructions?ws=nginx&os=ubuntufocal). Works wonderfully usually.

**Comment by aunefyren** [comment]
This is a badly named error that occurs when decrypting the cookie using Base64  for verification. My first guess is that maybe the line here:

`$cookie = htmlspecialchars($data->cookie)`;

On line 30 in wrapperr/api/validate_login_cookie.php is removing some important characters. Doubt it though, as I probably would have experienced it too. But, try changing it: 

`$cookie = $data->cookie`;

Also make sure that the value `token_encrypter` in your config JSON stays consistent, as this is used to encrypt and decrypt the cookies it supplies. 

Also make sure you are sending the whole value of the cookie given by the API.

**Comment by VP-EN** [comment]
So I edited the `validate_login_cookie.php` on line 30.
I checked the `config.json` for any changes to the `token_encrypter` throughout multiple logins, and nginx restarts. 
No changes to the token.

I've tried:
```
curl --header "Content-Type: application/json" \
  --request POST \
  --data '{"cookie":"w3fviZ0uEUE6gKAXmp6jp3Mncq9pxFefHuK5U5xGurDX"}' \
  https://wrapped.nscbox.com/api/validate_login_cookie.php
```

**Output:** (verbose) - it doesn't output anything if no `-v` applied to the curl command.
https://0bin.net/paste/-bjMl8o2#k-RDDXPtg//FCmjwDu1lk9iRa05ShF75DCoy4IOaG0/

and 

```
curl --header "Content-Type: application/json" \
  --request POST \
  --data '{"cookie":"plex-wrapped=w3fviZ0uEUE6gKAXmp6jp3Mncq9pxFefHuK5U5xGurDX"}' \
  https://wrapped.nscbox.com/api/validate_login_cookie.php
```

**Output:** (verbose)
https://0bin.net/paste/j6ZPC5T9#Cc8ndfkEncJGnhKqdnnDxyosV2vYyWkmocA1xZX90Iy

**Comment by aunefyren** [comment]
Did you see my edit and try that?

**Comment by VP-EN** [comment]
> Did you see my edit and try that?

Yes?
I think it's what I did here:
```
curl --header "Content-Type: application/json" \
  --request POST \
  --data '{"cookie":"w3fviZ0uEUE6gKAXmp6jp3Mncq9pxFefHuK5U5xGurDX"}' \
  https://wrapped.nscbox.com/api/validate_login_cookie.php
```

**Output:** (verbose) - it doesn't output anything if no `-v` applied to the curl command.
https://0bin.net/paste/-bjMl8o2#k-RDDXPtg//FCmjwDu1lk9iRa05ShF75DCoy4IOaG0/


removed the `plex-wrapped=` part. It just didn't output anything

**Comment by aunefyren** [comment]
> Edit
> Tried on another server of mine, using the docker image. On there it seems to work fine.
> Could it be because I'm using php8 instead of php7.4 or something?

Mmmm, that could be. Never tried php8 before. I'll attempt to install it on php8 later today and track down the issue

**Comment by aunefyren** [comment]
Also, thank you for the effort. Really helps making this project as good as it can be!

**Comment by VP-EN** [comment]
> > Edit
> > Tried on another server of mine, using the docker image. On there it seems to work fine.
> > Could it be because I'm using php8 instead of php7.4 or something?
> 
> Mmmm, that could be. Never tried php8 before. I'll attempt to install it on php8 later today and track down the issue

Thanks 👍 

> Also, thank you for the effort. Really helps making this project as good as it can be!

Huge thanks to you too! I really love these "small" additions to the "Plex-i-verse" so to speak.
So thanks for making it, and thanks for helping me sort out this issue 😄

**Comment by VP-EN** [comment]
Oh wow, I solved it I think.

I was missing some PHP module apparently. I purged my PHP8 install and installed from [this guide](https://www.linode.com/docs/guides/install-php-8-for-apache-and-nginx-on-ubuntu/) from Linode.
Including the modules:
`sudo apt install php-common php-mysql php-cgi php-mbstring php-curl php-gd php-xml php-xmlrpc php-pear`

Not sure which one of them helped, but it works now. Phew! It's always the small things, that end up causing the most trouble
